substrate/frame/nomination-pools/src/migration.rs

```diff
@@ -724,4 +724,103 @@ pub mod v5 {
 			Ok(())
 		}
 	}
+
+	/// This migration accumulates and initializes the [`TotalValueLocked`] for all pools.
+	pub struct VersionUncheckedMigrateV5ToV6<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for VersionUncheckedMigrateV5ToV6<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let migrated = BondedPools::<T>::count();
+			// The TVL should be the sum of all the funds that are actively staked and in the
+			// unbonding process of the account of each pool.
+			let tvl: BalanceOf<T> = BondedPools::<T>::iter()
+				.map(|(id, inner)| {
+					T::Staking::total_stake(
+						&BondedPool { id, inner: inner.clone() }.bonded_account(),
+					)
+					.unwrap_or_default()
+				})
+				.reduce(|acc, total_balance| acc + total_balance)
+				.unwrap_or_default();
+
+			TotalValueLocked::<T>::set(tvl);
+
+			log!(info, "Upgraded {} pools with a TVL of {:?}", migrated, tvl);
+
+			// reads: migrated * (BondedPools +  Staking::total_stake) + count + onchain
+			// version
+			//
+			// writes: current version + TVL
+			T::DbWeight::get().reads_writes(migrated.saturating_mul(2).saturating_add(2).into(), 2)
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+			ensure!(
+				PoolMembers::<T>::iter_keys().count() == PoolMembers::<T>::iter_values().count(),
+				"There are undecodable PoolMembers in storage. This migration will not fix that."
+			);
+			ensure!(
+				BondedPools::<T>::iter_keys().count() == BondedPools::<T>::iter_values().count(),
+				"There are undecodable BondedPools in storage. This migration will not fix that."
+			);
+			ensure!(
+				SubPoolsStorage::<T>::iter_keys().count() ==
+					SubPoolsStorage::<T>::iter_values().count(),
+				"There are undecodable SubPools in storage. This migration will not fix that."
+			);
+			ensure!(
+				Metadata::<T>::iter_keys().count() == Metadata::<T>::iter_values().count(),
+				"There are undecodable Metadata in storage. This migration will not fix that."
+			);
+
+			Ok(Vec::new())
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(_data: Vec<u8>) -> Result<(), TryRuntimeError> {
```

Comment:
I totally agree, but I do wonder how you wold check for that without repeating the exact same calculation. For further checking it's also possible to check for keeping the upper bound of the sum of all `total_balance` of all `PoolMembers` but thats also not an exact check as you seem to wish for. Iterating through all the PoolMembers is sadly very heavy. Can you give me an idea on how to properly do the `post_upgrade` checks?

Reply:
`post_upgrade` / `pre_upgrade` are not executed on-chain, so it doesn't matter if they're heavy.

you can abstract the tvl calculation into a function if you'd like to reuse it between `OnRuntimeUpgrade` methods

Reply:
thanks, done. Now both checks I mentioned are included in `post_upgrade`.

Reply:
This is probably not needed since try state checks already include them and we can run migration checks with the try-state checks which will verify again what you are doing in post_upgrade. I think just checking the storage item is set with a value is good enough and what we generally do in other migrations.